File: src/test/java/io/pivotal/security/controller/v1/RegenerateControllerTest.java
package io.pivotal.security.controller.v1;

import io.pivotal.security.CredentialManagerApp;
import io.pivotal.security.audit.EventAuditLogService;
import io.pivotal.security.audit.RequestUuid;
import io.pivotal.security.handler.RegenerateHandler;
import io.pivotal.security.util.DatabaseProfileResolver;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.test.mock.mockito.SpyBean;
import org.springframework.http.MediaType;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.context.WebApplicationContext;

import static io.pivotal.security.util.AuthConstants.UAA_OAUTH2_PASSWORD_GRANT_TOKEN;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.springframework.security.test.web.servlet.setup.SecurityMockMvcConfigurers.springSecurity;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@RunWith(SpringRunner.class)
@ActiveProfiles(value = "unit-test", resolver = DatabaseProfileResolver.class)
@SpringBootTest(classes = CredentialManagerApp.class)
@Transactional
public class RegenerateControllerTest {
  @Autowired
  private WebApplicationContext webApplicationContext;

  @MockBean
  private RegenerateHandler regenerateHandler;

  @SpyBean
  private EventAuditLogService eventAuditLogService;

  private MockMvc mockMvc;

  @Before
  public void beforeEach() {
    mockMvc = MockMvcBuilders
        .webAppContextSetup(webApplicationContext)
        .apply(springSecurity())
        .build();
  }

  @Test
  public void POST_regeneratesThePassword_andPersistsAnAuditEntry() throws Exception {
    mockMvc.perform(makeRegenerateRequest()).andDo(print()).andExpect(status().isOk());

    Mockito.verify(regenerateHandler).handleRegenerate(eq("picard"), any());
    Mockito.verify(eventAuditLogService).auditEvents(any(RequestUuid.class), any());
  }

  @Test
  public void POST_withSignedBy_regeneratesAllCertificatesSignedByCA_andPersistsAnAuditEntry() throws Exception {
    mockMvc.perform(makeBulkRegenerateRequest()).andDo(print()).andExpect(status().isOk());

    Mockito.verify(regenerateHandler).handleBulkRegenerate(eq("/some-ca"), any());
    Mockito.verify(eventAuditLogService).auditEvents(any(RequestUuid.class), any());
  }

  private MockHttpServletRequestBuilder makeRegenerateRequest() {
    return post("/api/v1/regenerate")
        .header("Authorization", "Bearer " + UAA_OAUTH2_PASSWORD_GRANT_TOKEN)
        .accept(MediaType.APPLICATION_JSON)
        .contentType(MediaType.APPLICATION_JSON)
        .content("{\n"
            + "  \"name\" : \"picard\"\n"
            + "}");
  }

  private MockHttpServletRequestBuilder makeBulkRegenerateRequest() {
    return post("/api/v1/bulk-regenerate")
        .header("Authorization", "Bearer " + UAA_OAUTH2_PASSWORD_GRANT_TOKEN)
        .accept(MediaType.APPLICATION_JSON)
        .contentType(MediaType.APPLICATION_JSON)
        .content("{\n"
            + "  \"signed_by\" : \"/some-ca\"\n"
            + "}");
  }
}

File: backends/credhub/src/test/kotlin/org/cloudfoundry/credhub/controllers/v1/regenerate/RegenerateControllerTest.kt
package org.cloudfoundry.credhub.controllers.v1.regenerate

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import org.bouncycastle.jcajce.provider.BouncyCastleFipsProvider
import org.cloudfoundry.credhub.constants.CredentialType
import org.cloudfoundry.credhub.credential.StringCredentialValue
import org.cloudfoundry.credhub.helpers.CredHubRestDocs
import org.cloudfoundry.credhub.helpers.MockMvcFactory
import org.cloudfoundry.credhub.helpers.credHubAuthHeader
import org.cloudfoundry.credhub.regenerate.RegenerateController
import org.cloudfoundry.credhub.utils.BouncyCastleFipsConfigurer
import org.cloudfoundry.credhub.views.BulkRegenerateResults
import org.cloudfoundry.credhub.views.CredentialView
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.skyscreamer.jsonassert.JSONAssert
import org.springframework.http.MediaType
import org.springframework.restdocs.JUnitRestDocumentation
import org.springframework.restdocs.mockmvc.MockMvcRestDocumentation.document
import org.springframework.restdocs.payload.JsonFieldType
import org.springframework.restdocs.payload.PayloadDocumentation
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.security.Security
import java.time.Instant
import java.util.UUID

@RunWith(SpringRunner::class)
class RegenerateControllerTest {
    @Rule
    @JvmField
    val restDocumentation = JUnitRestDocumentation()

    private lateinit var spyRegenerateHandler: SpyRegenerateHandler
    private lateinit var mockMvc: MockMvc

    private val objectMapper: ObjectMapper = ObjectMapper()
    lateinit var metadata: JsonNode

    companion object {
        @BeforeClass
        @JvmStatic
        fun setUpAll() {
            BouncyCastleFipsConfigurer.configure()
        }
    }

    @Before
    fun beforeEach() {
        spyRegenerateHandler = SpyRegenerateHandler()
        val regenerateController = RegenerateController(spyRegenerateHandler)

        metadata = objectMapper.readTree("{\"description\":\"example metadata\"}")

        mockMvc = MockMvcFactory.newSpringRestDocMockMvc(regenerateController, restDocumentation)
        if (Security.getProvider(BouncyCastleFipsProvider.PROVIDER_NAME) == null) {
            Security.addProvider(BouncyCastleFipsProvider())
        }
    }

    @Test
    fun postRegenerateReturnsResults() {
        val randomUUID = UUID.randomUUID()
        spyRegenerateHandler.handleregenerateReturnsCredentialview =
            CredentialView(
                Instant.ofEpochSecond(1549053472L),
                randomUUID,
                "/some-name",
                CredentialType.VALUE.type.lowercase(),
                metadata,
                StringCredentialValue("some-value"),
            )

        val actualResponse =
            mockMvc
                .perform(
                    post("/api/v1/regenerate")
                        .credHubAuthHeader()
                        .accept(MediaType.APPLICATION_JSON)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(
                            """
                            {
                                "name": "/some-name",
                                "metadata": { "some": "example metadata"}
                            }
                            """.trimIndent(),
                        ),
                ).andExpect(status().isOk)
                .andDo(
                    document(
                        CredHubRestDocs.DOCUMENT_IDENTIFIER,
                        PayloadDocumentation.requestFields(
                            PayloadDocumentation
                                .fieldWithPath("name")
                                .description("The credential name to regenerate.")
                                .type(JsonFieldType.STRING),
                            PayloadDocumentation
                                .fieldWithPath("metadata")
                                .description("Additional metadata of the credential.")
                                .optional(),
                            PayloadDocumentation
                                .fieldWithPath("metadata.*")
                                .ignored(),
                        ),
                    ),
                ).andReturn()
                .response

        // language=json
        val expectedResponse =
            """
            {
              "type": "value",
              "version_created_at": "2019-02-01T20:37:52Z",
              "id": "$randomUUID",
              "name": "/some-name",
              "metadata": { "description": "example metadata"},
              "value": "some-value"
            }
            """.trimIndent()

        JSONAssert.assertEquals(expectedResponse, actualResponse.contentAsString, true)

        assertThat(spyRegenerateHandler.handleRegenerateCalledWithCredentialName, equalTo("/some-name"))
    }

    @Test
    fun postBulkregenerateReturnsResults() {
        spyRegenerateHandler.handleBulkRegenerateReturnsBulkRegenerateResults =
            {
                val bulkRegenerateResults = BulkRegenerateResults()
                bulkRegenerateResults.setRegeneratedCredentials(
                    mutableSetOf(
                        "/some-credential-name",
                        "/some-other-credential-name",
                    ),
                )

                bulkRegenerateResults
            }()

        val actualResponse =
            mockMvc
                .perform(
                    post("/api/v1/bulk-regenerate")
                        .credHubAuthHeader()
                        .accept(MediaType.APPLICATION_JSON)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(
                            """
                            {
                                "signed_by": "/some-ca"
                            }
                            """.trimIndent(),
                        ),
                ).andExpect(status().isOk)
                .andDo(
                    document(
                        CredHubRestDocs.DOCUMENT_IDENTIFIER,
                        PayloadDocumentation.requestFields(
                            PayloadDocumentation
                                .fieldWithPath("signed_by")
                                .description("The name of the CA that signs regenerated certificates.")
                                .type(JsonFieldType.STRING),
                        ),
                    ),
                ).andReturn()
                .response

        // language=json
        val expectedResponse =
            """
            {
              "regenerated_credentials": [
                "/some-credential-name",
                "/some-other-credential-name"
              ]
            }
            """.trimIndent()

        JSONAssert.assertEquals(expectedResponse, actualResponse.contentAsString, true)

        assertThat(spyRegenerateHandler.handleBulkRegenerateCalledWithSignerName, equalTo("/some-ca"))
    }
}
